### PR TITLE
Add settings storage capabilities for users

### LINKED ISSUE #286: Add `Settings` storage capabilities for `User`
# User Settings

Users will be able to have configurable settings that should persist.

# Implementation

I believe this should be under a new `settings` table which is tied to the `user` entity. For now, I think the following schema would be sufficient:

```
- id
- user_id
- default_ruleset
```

# Objectives

- [x] Create `settings` table that is tied to a user
- [x] Create associated service and repository
- [ ] Create a settings controller
- [x] Allow users to modify their own settings similar to the `me` controller (the route could even be `/me/settings` if we want, @myssto let me know your thoughts).
- [x] Create endpoint for modifying one's own default ruleset. e.g. `PATCH /settings` or `PATCH /me/settings`

### PR DESCRIPTION
**Pull request by myssto**
Adds a `UserSettings` entity that will hold user controllable settings for `otr-web` going forward. Currently holds two values, `Ruleset DefaultRuleset` and `bool DefaultRulesetIsControlled`. The rationale is that by default, users will have their preferred ruleset synced with the ruleset they have selected on their osu! profile. If they wish to have a different preferred ruleset for otr, they can change it, and the value will no longer sync with their osu! profile.

UserSettings are populated with defaults via a migration. By default, all users' rulesets will be set to and synced with the one selected on their osu! profile.

# New Endpoints
- `POST` `/me/settings/ruleset`
  - Updates the ruleset for the currently logged in user
  - Takes a request body of Ruleset
  - Example (Updates user ruleset to `Ruleset.Catch`)
    ```js
    fetch("otr.stagec.xyz/api/v1/me/settings/ruleset",
      {
        method: "POST",
        headers: new Headers().append("Authorization", "Bearer {}"),
        body: JSON.stringify(2),
        redirect: "follow"
      });
    ````
- `POST` `/me/settings/ruleset:sync`
  - Syncs the ruleset for the currently logged in user to their osu! profile ruleset
  - Example
    ```js
    fetch("otr.stagec.xyz/api/v1/me/settings/ruleset:sync",
      {
        method: "POST",
        headers: new Headers().append("Authorization", "Bearer {}"),
        redirect: "follow"
      });
    ````
- `POST` `/users/{id}/settings/ruleset`
  - Updates the ruleset for a user
  - Admin and owner access only
  - Takes a request body of Ruleset
  - Example (Updates ruleset to `Ruleset.Catch` for user `703` (Myssto))
    ```js
    fetch("otr.stagec.xyz/api/v1/users/703/settings/ruleset",
      {
        method: "POST",
        headers: new Headers().append("Authorization", "Bearer {}"),
        body: JSON.stringify(2),
        redirect: "follow"
      });
    ````
- `POST` `/users/{id}/settings/ruleset:sync`
  - Syncs the ruleset a user to their osu! profile ruleset
  - Example (Syncs ruleset for user `703` (Myssto))
    ```js
    fetch("otr.stagec.xyz/api/v1/users/703/settings/ruleset:sync",
      {
        method: "POST",
        headers: new Headers().append("Authorization", "Bearer {}"),
        redirect: "follow"
      });
    ````

# Schema Changes
- `UserDTO` (Returns from `GET` `/me` and `GET` `/users/{id}`)
  - Removed `ruleset`
  - Added `settings`
  - ```json
    {
    "id": 703,
    "scopes": [
        "whitelist"
    ],
    "playerId": 7470,
    "osuId": 18068913,
    "country": "US",
    "username": "rimjob",
    "settings": {
        "ruleset": 0,
        "rulesetIsControlled": false
        }
    }
    ```

Closes #286

### CODE REVIEW

File: API/Entities/UserSettings.cs [hunk 1]
```diff
@@ -0,0 +1,50 @@
...
+    /// The id of the associated user
+    /// </summary>
+    [Column("user_id")]
+    public int UserId { get; set; }
+
+    /// <summary>
+    /// The default ruleset for the associated user
+    /// </summary>
+    [Column("default_ruleset")]
+    public Ruleset? DefaultRuleset { get; set; }
```
Comment (hburn7) [inline API/Entities/UserSettings.cs:34]: nitpick, but I prefer the Created/Updated props to be on the bottom (I would even undo and recreate the migration):

From top to bottom:
- Id
- Properties
- Created
- Updated
- Navigation props / collections

File: API/Repositories/Implementations/UserRepository.cs [hunk 2]
```diff
@@ -7,15 +7,33 @@ namespace API.Repositories.Implementations;
...
-public class UserRepository(OtrContext context) : RepositoryBase<User>(context), IUserRepository
+public class UserRepository(OtrContext context, IUserSettingsRepository userSettingsRepository) : RepositoryBase<User>(context), IUserRepository
 {
     private readonly OtrContext _context = context;
 
+    public override async Task<User> CreateAsync(User entity)
+    {
+        if (!entity.PlayerId.HasValue)
+        {
+            throw new NullReferenceException("Attempting to create a User entity without a PlayerId");
```
Comment (hburn7) [inline API/Repositories/Implementations/UserRepository.cs:18]: Past-tense would be better

File: API/Services/Implementations/UserSettingsService.cs [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    public async Task<bool> SyncRulesetAsync(int userId)
+    {
+        User? user = await userRepository.GetAsync(userId);
+
+        if (user is null)
+        {
+            return false;
+        }
+
+        user.Settings.DefaultRuleset = user.Player.Ruleset ?? Ruleset.Standard;
```
Comment (hburn7) [inline API/Services/Implementations/UserSettingsService.cs:35]: I smell a nullreference here, I'd ensure user settings isn't null.

**Review by hburn7 — CHANGES_REQUESTED** [review]
Have not tested locally yet, this is just a code review.

**Comment by hburn7** [comment]
Schema changes need to be implemented in web implementation before merging.

**Comment by myssto** [comment]
Tested locally against otr-web#189 . Everything looks good to me.

**Review by hburn7 — APPROVED** [review]

**Pull request merged**